### PR TITLE
Add e2e test for invalid PSBT upload

### LINKED ISSUE #442: End-to-End Testing Coverage
## Overview

End-to-end testing is an essential part of ensuring Caravan works correctly from a user's perspective. Our e2e tests simulate real user workflows by interacting with the actual application interface, coordinating with a Bitcoin regtest node, and verifying that critical flows work as expected.

This is an excellent area for new contributors to get familiar with Caravan since it involves working with actual user flows without requiring deep knowledge of the internal architecture. Each test scenario helps you understand how Caravan works while improving our test coverage.

## How E2E Testing Works in Caravan

Our e2e tests use Playwright to automate browser interactions with the Caravan Coordinator application. When you run the tests, the system automatically:

1. Spins up a Bitcoin regtest node in Docker
2. Launches the Caravan application 
3. Creates test wallets with funded addresses
4. Runs automated tests that click buttons, fill forms, and verify outcomes
5. Cleans up all resources when finished

Everything runs in an isolated environment, so you can safely test transaction flows without touching real Bitcoin.

## Current Test Coverage

We currently have three main test suites covering core workflows:

### 01-multisig_setup.spec.ts
Tests the wallet creation and configuration flow:
- Loading the Caravan homepage
- Navigating through wallet setup
- Testing Bitcoin node connections with various scenarios (success, wrong URL, incorrect credentials)
- Creating a 2-of-3 multisig wallet
- Entering extended public keys from test wallets
- Downloading wallet configuration files

### 02-wallet_regtest.spec.ts
Tests wallet configuration and address management:
- Modifying wallet configurations for regtest network
- Extracting and setting proper BIP32 paths and fingerprints
- Importing wallets into Caravan
- Importing addresses to Bitcoin node
- Verifying address generation and display
- Address pagination and navigation

### 03-transaction_flow.spec.ts
Tests transaction creation and broadcasting:
- Creating transactions with automatic coin selection
- Creating transactions with manual UTXO selection
- Downloading unsigned PSBTs
- Signing PSBTs with multiple wallets
- Uploading signatures (both as text and file upload)
- Broadcasting transactions
- Verifying successful transaction broadcast

### Prerequisites
- Docker installed and running
- Node.js (check `.nvmrc` for version)

Also you'd need to change resource setting of Docker to allow atleast 8gbs of ram access

Image

### Getting Started

1. **Clone and navigate to the project:**
   ```bash
   git clone https://github.com/caravan-bitcoin/caravan.git
   cd caravan
   ```

2. **Install dependencies:**
   ```bash
   npm install
   cd apps/coordinator
   npm run test:e2e:install  # Installs Playwright browsers
   ```

3. **Set up environment:**
   ```bash
   cd e2e
   cp .env.example .env
   ```

4. **Run the tests:**
   ```bash
   # From project root
   npm run test:e2e --filter=caravan-coordinator
   
   # Or from coordinator directory
   cd apps/coordinator
   npm run test:e2e          # Run all tests
   npm run test:e2e:ui       # Interactive UI mode
   npm run test:e2e:headed   # See the browser
   ```

The test infrastructure handles everything automatically - no need to manually start Docker containers or configure the Bitcoin node.

### Reliability of existing tests

In addition to the below tests to build out, there is some flakiness that exists in the current tests where timeouts will eventually be hit. Our tests will run retries and if after 3 retries they still fail then the suite will fail. Recreating these failures and trying to improve the performance and reliability would be an important contribution to our codebase.

Here are workflows we'd like to add e2e coverage for. Each represents a small, focused PR opportunity:

### Wallet Management
- [ ] Importing an existing wallet configuration file
- [ ] Testing different address types (P2WSH, P2SH-P2WSH, P2TR)
- [ ] Verifying wallet with different quorum configurations (1-of-2, 3-of-5)
- [ ] Testing wallet name editing
- [ ] Exporting wallet descriptors

### Address Operations
- [ ] Generating and displaying change addresses
- [ ] Testing address verification with QR codes
- [ ] Batch address generation
- [ ] Address reuse warnings

### Transaction Workflows
- [ ] PSBT management (downloading, uploading, importing with signatures)
- [ ] Creating transactions to multiple recipients
- [ ] RBF (Replace-By-Fee) transaction creation
- [ ] CPFP (Child-Pays-For-Parent) transaction creation
- [ ] Testing different fee rate selections
- [ ] Transaction preview and review flow
- [ ] Canceling transaction creation mid-flow

### Error Handling
- [ ] Invalid address entry
- [ ] Insufficient balance scenarios
- [ ] Invalid PSBT upload
- [ ] Signature mismatch errors
- [ ] Network disconnection handling
- [ ] Invalid fee rate inputs

### Hardware Wallet Integration
- [ ] Coldcard PSBT import/export flow
- [ ] Testing different signing device combinations
- [ ] Verifying proper input ordering

### Client Connections
- [ ] Testing different public API providers
- [ ] Private node connection failure recovery
- [ ] Network switching (testnet/regtest)

### 1. Pick a Test Scenario
Choose one of the unchecked items from the list above, or propose a new one by commenting on this issue. 

IMPORTANT: Choose just one per PER to make them easier to review and validate.

### 2. Set Up Your Development Environment
Follow the setup instructions above to get e2e tests running locally.

### 3. Write Your Test
Create a new test file or add to an existing one in `apps/coordinator/e2e/tests/`. Look at the existing tests as examples. Each test should:
- Have a clear, descriptive name
- Focus on a single workflow or scenario
- Use the shared test utilities in `e2e/utils/` and `e2e/testhelpers/`
- Clean up after itself (handled by global teardown)

### 4. Test Locally
Run your test multiple times to ensure it's stable:
```bash
npm run test:e2e:headed  # Watch it run
npm run test:e2e         # Verify it passes consistently
```

### 5. Submit a Pull Request
Create a PR with:
- A descriptive title like "e2e: Add test for RBF transaction creation"
- A brief description of what the test covers
- Reference to this issue
- Any setup notes or edge cases discovered

## Example Test Structure

```typescript
test.describe("Your Test Suite Name", () => {
  test.beforeEach(async ({ page }) => {
    // Navigate to starting point
    await page.goto("/#/wallet");
    // Load wallet, connect to node, etc.
  });

  test("should perform the workflow", async ({ page }) => {
    // Your test steps here
    // Use page.locator() to find elements
    // Use expect() to verify outcomes
  });
});
```

## Helpful Resources

- **Playwright Documentation**: https://playwright.dev/docs/intro
- **Existing Test Examples**: `apps/coordinator/e2e/tests/`
- **Test Helpers**: `apps/coordinator/e2e/testhelpers/` - Utilities for common operations
- **Bitcoin RPC Client**: `apps/coordinator/e2e/utils/bitcoinClient.ts` - For blockchain operations

## Questions or Need Help?

Feel free to comment on this issue or join our discord channel or here on the issue . We're here to help new contributors get started with testing and learn about Caravan along the way :) 

---

*Note: A demo video showing how to run and write e2e tests is added below .* ( I did it for mac if anyone does it on windows/linux , please feel free to share here

### PR DESCRIPTION
**Pull request by harrshita123**
Test / Quality improvement (E2E test)

Fixes #442

This PR adds an end-to-end test to validate Caravan’s behavior when an invalid PSBT file is uploaded.

The test ensures that:
- Uploading an invalid PSBT triggers a visible error alert
- The transaction signing flow is blocked for invalid PSBT inputs

This improves test coverage and prevents regressions related to PSBT validation.

- [x] I have tested my changes thoroughly.
- [x] I have added tests to cover my changes.
- [x] I have run the test suite locally, and all tests pass.
- [x] I have written tests for all new changes/features.
- [x] I have followed the project's coding style and conventions.

- Added E2E test in:
  `apps/coordinator/e2e/tests/03-transaction_flow.spec.ts`
- Verified locally using `npm run test:e2e`

 Have you read the contributing guide?
Yes

### CODE REVIEW

File: .changeset/invalid-psbt-e2e-test.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+"caravan-coordinator": none
```
Comment (Legend101Zz) [inline .changeset/invalid-psbt-e2e-test.md]: actually we don't need a changeset for this , changeset it for packages and coordinator version , we don't it for e2e test suite ... @bucko13 please correct me if I am wrong here :)

**Review by Legend101Zz — APPROVED** [review]
Just a nit , rest looks good to me thanks for the PR , can you also please attach a screen shot showing the playwright running the test , would be really helpful thanks

**Comment by harrshita123** [comment]
Screenshot 2026-02-17 at 10 34 41 PM

File: .changeset/invalid-psbt-e2e-test.md (hunk 1 above)
Comment (harrshita123) [reply .changeset/invalid-psbt-e2e-test.md]: > > actually we don't need a changeset for this , changeset it for packages and coordinator version , we don't it for e2e test suite ... @bucko13 please correct me if I am wrong here :)
> 
@Legend101Zz 
> Thanks for the review and the helpful suggestion!
> 
> You are absolutely right . since this change only adds an end-to-end test and does not modify any package code or affect the coordinator version, a changeset isn’t required. I have removed the changeset file accordingly.
> 
> I have also attached a screenshot of the Playwright test suite running locally (npm run test:e2e) showing all tests passing, including the newly added invalid PSBT test.
> 
> please let me know if there is anything else I should improve!

**Review by Legend101Zz — APPROVED** [review]
LGTM

**Comment by bucko13** [comment]
With #487 we need to update to the new e2e system. will require resolving some merge conflicts.

**Comment by Legend101Zz** [comment]
Hi @harrshita123  Thanks for your patience. As Buck suggested above due to the E2E changes the testsiote was updated , Could you please pull in the latest changes and review the updated E2E setup we’ve implemented, then adjust your changes accordingly?

**Comment by harrshita123** [comment]
> Hi @harrshita123 Thanks for your patience. As Buck suggested above due to the E2E changes the testsiote was updated , Could you please pull in the latest changes and review the updated E2E setup we’ve implemented, then adjust your changes accordingly?

 I will sync with the latest main, rebase this PR, and update my changes to align with the  new E2E setup from #487. I will take care of the merge conflicts and push an update soon.

**Comment by harrshita123** [comment]
Implemented the updates based on your guidance for the new E2E setup and resolved the related conflicts. I also re- ran the E2E suite and all tests are passing (15 passed). Sharing the screenshot of the passing run here for reference.

Screenshot 2026-03-11 at 10 44 03 PM

**Review by bucko13 — APPROVED** [review]

**Comment by harrshita123** [comment]
Should i have to add changesets ?

File: apps/coordinator/e2e/tests/mutate/transaction-flow.spec.ts [hunk 2]
```diff
@@ -44,6 +44,21 @@ test.describe.serial("Transaction Creation and Signing", () => {
...
+  test("invalid PSBT upload: shows error and blocks signing", async ({
+    sendTab,
+    page,
+  }) => {
+    const invalidPsbtPath = path.join(uploadDir, "invalid.psbt");
+    fs.writeFileSync(invalidPsbtPath, "this-is-not-a-valid-psbt");
+
+    await sendTab.importPsbtFile(invalidPsbtPath);
+
+    await expect(page.getByRole("alert")).toBeVisible({ timeout: 10000 });
```
Comment (Legend101Zz) [inline apps/coordinator/e2e/tests/mutate/transaction-flow.spec.ts:56]: The assertions at the end are leaking raw Playwright locators into the test body — this should go through the page object instead. `sendTab` is right there; something like `sendTab.expectImportError()` (or whatever makes sense on the POM) would keep the test readable and keep selector details out of spec files.

So can you plaese move the `page.getByRole("alert")` and sign-button assertions behind a POM method on `sendTab`.

As if we don't follow the POM model we can again go back to having flaky tests :)

**Review by Legend101Zz — CHANGES_REQUESTED** [review]
@harrshita123 thanks for the PR , I had a comment on moving all the assertion logic to pages , you can read more in the contribution.md to understand how the new e2e architecture works :)

**Comment by harrshita123** [comment]
@bucko13 
Please open this pull request .

**Comment by bucko13** [comment]
Hey you should be able to re-open yourself. It's closed just with automation to keep the repo from getting overrun with stale PRs. When youre ready to pick it back up feel free to re-open as you're ready

**Comment by harrshita123** [comment]
@bucko13 I tried reopening this PR, but GitHub is not allowing me to reopen it from my account. Could you please reopen and review it when you have a chance?

**Comment by harrshita123** [comment]
I tried reopening this PR, but GitHub is not allowing me to reopen it from my account. Could you please reopen and review it when you have a chance?
